### PR TITLE
search: configure query parser for handling special characters

### LINKED ISSUE #2930: search: exact dataset title
Revise tokenisers and analysers so that exact searches would work.

For example, for the dataset http://opendata.cern.ch/record/1, the following query does not give anything:

```
/BTau/Run2010B-Apr21ReReco-v1/AOD
```

even though this is exact match with record title.

It even leads to *Error: Invalid query", most probably due to the use of slashes.

The users have to use double quotes:

```
"/BTau/Run2010B-Apr21ReReco-v1/AOD"
```

which finds the record.

However, forcing the use of double quotes is not very user-friendly.

Let's see if we can amend Elasticsearch mappings and analysers so that exact matches will be returned directly.

### LINKED ISSUE #3042: search: configure query parser internally for handling special characters
Using Luqum -> https://luqum.readthedocs.io/en/latest/about.html

### PR DESCRIPTION
**Pull request by ParthS007**
closes #3042 
closes #2930

### CODE REVIEW

File: cernopendata/config.py [hunk 1]
```diff
@@ -281,7 +281,7 @@
             default_order='asc',
             order=2,
         ),
-        'title': dict(fields=['title.keyword'],
+        'title': dict(fields=['title.exact'],
```
Comment (tiborsimko) [inline cernopendata/config.py:284]: Sorting is well fixed :+1: Special characters appear in the front, which is OK for now, but we can also see later whether we could classify them into the alphabet, e.g. `/Btau` amongst `B`s rather than in front of the search results.

File: cernopendata/modules/records/search/query.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    :param self: REST view
+    :param search: Elastic search DSL search instance
+
+    :return: Tuple with search instance and URL arguments
+    """
+    def _default_parser(_query_string=None):
+        """Default parser that uses the Q() from elasticsearch_dsl."""
+        if '"' not in _query_string:
+            _query_string = _query_string.replace("/", "//")
```
Comment (tiborsimko) [inline cernopendata/modules/records/search/query.py:45]: This fixes the internal error for regexp-like queries, but it also leads to some unexpected findings, such as:

- searching for `10.7483/OPENDATA.9S5F.BY3B` will find `/glossary/Barn` as one of the matching hits.  Do you know why was this matched?  

- searching for `"10.7483/OPENDATA.9S5F.BY3B"` will find hundreds of results which are not related, like `record/349`.

An alternative approach could be to quote expressions... We can see IRL

Comment (tiborsimko) [reply cernopendata/modules/records/search/query.py:45]: Here is a table showing an overview of the above findings:

query | opendata.cern.ch (production) | opendata-dev.cern.ch (master) | PR 3054
----- | ----------------------------- | ----------------------------- | -------
10.7483/OPENDATA.9S5F.BY3B | error | 357 results | 359 results
"10.7483/OPENDATA.9S5F.BY3B" | 1 result | error | 5444 results 
10.7483 | 2123 results | 357 results | 359 results 

Basically the PROD system has the best result, the only thing we want to correct is the "error"  when people don't use quotes.  The PR fixes the error, but finds too much hits for quoted query, and not enough hits for partial query.

Comment (ParthS007) [reply cernopendata/modules/records/search/query.py:45]: Yes, I am currently testing this one.

Comment (tiborsimko) [reply cernopendata/modules/records/search/query.py:45]: Ditto for CMS dataset title search; while the PR fixes the error situations, it also finds way too many "unrelated" results:

query | opendata.cern.ch (production) | opendata-dev.cern.ch (master) | PR 3054
----- | ----------------------------- | ----------------------------- | -------
/BTau/Run2010B-Apr21ReReco-v1/AOD | error | error | 4851 results
"/BTau/Run2010B-Apr21ReReco-v1/AOD" | 2 results | error | 4851 results 
BTau | 5 results | 5 results | 5 results

File: tests/test_cernopendata_query_parser.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of CERN Open Data Portal.
+# Copyright (C) 2017 CERN.
```
Comment (tiborsimko) [inline tests/test_cernopendata_query_parser.py:4]: s/2017/2021/g ?

File: cernopendata/modules/records/search/query.py (hunk 2 above)
Comment (ParthS007) [reply cernopendata/modules/records/search/query.py]: Ping @tiborsimko

**Review by tiborsimko — APPROVED** [review]
